src/theory/quantifiers/instantiate.cpp

```diff
@@ -0,0 +1,831 @@
+    }
+  }
+  else
+  {
+    for (std::pair<const Node, inst::InstMatchTrie>& t : d_inst_match_trie)
+    {
+      getInstantiationTermVectors(t.first, insts[t.first]);
+    }
+  }
+}
+
+void Instantiate::getExplanationForInstLemmas(
+    const std::vector<Node>& lems,
+    std::map<Node, Node>& quant,
+    std::map<Node, std::vector<Node> >& tvec)
+{
+  if (options::trackInstLemmas())
+  {
+    if (options::incrementalSolving())
+    {
+      for (std::pair<const Node, inst::CDInstMatchTrie*>& t :
+           d_c_inst_match_trie)
+      {
+        t.second->getExplanationForInstLemmas(t.first, lems, quant, tvec);
+      }
+    }
+    else
+    {
+      for (std::pair<const Node, inst::InstMatchTrie>& t : d_inst_match_trie)
+      {
+        t.second.getExplanationForInstLemmas(t.first, lems, quant, tvec);
+      }
+    }
+#ifdef CVC4_ASSERTIONS
+    for (unsigned j = 0; j < lems.size(); j++)
+    {
+      Assert(quant.find(lems[j]) != quant.end());
+      Assert(tvec.find(lems[j]) != tvec.end());
+    }
+#endif
+  }
+  Assert(false);
+}
+
+void Instantiate::getInstantiations(std::map<Node, std::vector<Node> >& insts)
+{
+  bool useUnsatCore = false;
+  std::vector<Node> active_lemmas;
+  if (options::trackInstLemmas() && getUnsatCoreLemmas(active_lemmas))
+  {
+    useUnsatCore = true;
+  }
+
+  if (options::incrementalSolving())
+  {
+    for (std::pair<const Node, inst::CDInstMatchTrie*>& t : d_c_inst_match_trie)
+    {
+      t.second->getInstantiations(
+          insts[t.first], t.first, d_qe, useUnsatCore, active_lemmas);
+    }
+  }
+  else
+  {
+    for (std::pair<const Node, inst::InstMatchTrie>& t : d_inst_match_trie)
+    {
+      t.second.getInstantiations(
+          insts[t.first], t.first, d_qe, useUnsatCore, active_lemmas);
+    }
+  }
+}
+
+void Instantiate::getInstantiations(Node q, std::vector<Node>& insts)
+{
+  if (options::incrementalSolving())
+  {
+    std::map<Node, inst::CDInstMatchTrie*>::iterator it =
+        d_c_inst_match_trie.find(q);
+    if (it != d_c_inst_match_trie.end())
+    {
+      std::vector<Node> active_lemmas;
+      it->second->getInstantiations(
+          insts, it->first, d_qe, false, active_lemmas);
+    }
+  }
+  else
+  {
+    std::map<Node, inst::InstMatchTrie>::iterator it =
+        d_inst_match_trie.find(q);
+    if (it != d_inst_match_trie.end())
+    {
+      std::vector<Node> active_lemmas;
+      it->second.getInstantiations(
+          insts, it->first, d_qe, false, active_lemmas);
+    }
+  }
+}
+
+Node Instantiate::getInstantiatedConjunction(Node q)
+{
+  Assert(q.getKind() == FORALL);
+  std::vector<Node> insts;
+  getInstantiations(q, insts);
+  if (insts.empty())
+  {
+    return NodeManager::currentNM()->mkConst(true);
+  }
+  else
+  {
+    Node ret;
+    if (insts.size() == 1)
+    {
+      ret = insts[0];
+    }
+    else
+    {
+      ret = NodeManager::currentNM()->mkNode(kind::AND, insts);
+    }
+    // have to remove q, TODO: avoid this in a better way
```

Comment:
add an issue?

Reply:
this was copied from previous code
I've removed the TODO but kept the comment since I don't think this is critical.